### ISSUE TITLE
Formalize the Master-Part relationship. 

### ISSUE THREAD

**Issue by dimitri-yatsenko** [datajoint/datajoint-matlab#47]
On the Python side, we have formalized the concept of a master-part relationship.  Let's mirror this formalism in the MATLAB implementation.

In a master-part relationship, there is a master table and one or more part tables.  We have formerly referred to part tables as "subtables".   Tuples in the master table always appear together with all their matching tuples in the part tables.  This is accomplished by disabling independent populate and delete methods of the part tables.  The part tables must be populated from the master table's makeTuples call as part of one transaction.  The master-part relationship enables dependencies on the master tuples and their matching parts, together.

In Python, part tables subclass `dj.Part` and are declared as nested classes of their master.  A part's internal table name has the form `master_table__part_table` with double underscores separating the master table's name from the part table's name.

MATLAB does not allow nested classes. Therefore, I propose the following solution that will keep the implementation compatible across the two languages:
1. The tier of a part table will be indicated as `part of schema.MasterClass` instead of the conventional `manual`, `lookup`, `imported`, and 'computed`.
2. Just like in Python, the table name will be `master_table__part_table`, where `part_table` is derived from the part class name.  To allow avoiding name collisions between identically named part tables of different master tables, the part table class can be prefixed with the master table class and it will be stripped internally.

Example (Python):

``` python
@schema
class Geometry(dj.Imported):
    definition = """  # probe geometry
        -> Design
        ----
        timestamp=CURRENT_TIMESTAMP : timestamp  # automatic timestamp
        """

    class EmitterLocation(dj.Part):
        definition = """  # 3D positions of emitters
            -> Geometry
            emitter :smallint
            ----
            emitter_x   :float  # (um)
            emitter_y   :float  # (um)
            emitter_z   :float  # (um)
            """

    class DetectorLocation(dj.Part):
        definition = """ # 3D position of detectors
            -> Geometry
            detector :smallint
            ----
            detector_x   :float   # (um)
            detector_y   :float   # (um)
            detector_z   :float   # (um)
            """
```

Equivalent MATLAB:

``` matlab
%{
probe.Geometry (imported)  # probe geometry
-> probe.Design
----
timestamp=CURRENT_TIMESTAMP : timestamp  # automatic timestamp
%}
classdef Geometry < dj.Relvar
begin
```

``` matlab
%{
probe.EmitterLocation (part of probe.Geometry)  # 3D positions of emitters
-> probe.Geometry
emitter :smallint
----
emitter_x   :float  # (um)
emitter_y   :float  # (um)
emitter_z   :float  # (um)
%}
classdef EmitterLocation < dj.Relvar
end
```

``` matlab
%{
probe.DetectorLocation (part of probe.Geometry)  # 3D positions of detectors
-> probe.Geometry
detector :smallint
----
detector_x   :float  # (um)
detector_y   :float  # (um)
detector_z   :float  # (um)
%}
classdef DetectorLocation < dj.Relvar
end
```

**Comment by aecker** [comment]
I agree to formalizing the master/part relationship. However, I suggest simplifying the complex syntax `part of schema.Class`. Instead, let's use only `part` and define the first reference (i.e. `-> schema.Class`) as the master.

**Comment by dimitri-yatsenko** [comment]
@aecker, I have considered that option but it seems like another implicit rule that must be explained in the documentation.   I am not sure if there are cases when users might not want to have the first reference to be to the master table.  I have not yet thought of such an example but I am thinking of its potential uses.

**Comment by fabiansinz** [comment]
I agree with Dimitri on that. Such kind of implicit rules can potentially lead to obscure behaviour. I don't think that `part of ...` is really a complex syntax and it makes it very clear what is happening.

**Comment by dimitri-yatsenko** [comment]
We can throw a warning if the first dependency is not on the master table.

**Comment by aecker** [comment]
It just feels unnecessarily redundant. In principle the entire first line of the definition is unnecessary; we could just subclass from `dj.Imported`, `dj.Part` etc. as in Python. So when we're changing the syntax anyway, I'd rather reduce redundancy that increase it. Ideally we use every name exactly once.

What would be a use case where the first reference is not the master? The order of references does not matter anyway, does it?

What kind of obscure behavior do you expect? From what I can tell the master/part relationship does not produce any "behavior". The only potential behavior I can see is preventing it from implementing `makeTuples()`, but the only way I can see we can accomplish that in Matlab is by defining the table their by subclassing `dj.Part` as in Python and not deriving `dj.Part` from `dj.AutoPopulate`.

**Comment by dimitri-yatsenko** [comment]
Yes, we could subclass from new classes `dj.Manual`, and `dj.Lookup`, `dj.Imported`, `dj.Computed`, etc.  However, this would introduce backward incompatibility.  

Master/part relationship has consequences for deletions.  Our current plan is to delete the master if any of the parts is deleted. This prevents deleting any parts independently.

Another implication is for naming conventions.  On the Python side, the convention is `schema.master_name__part_name` with double underscore separating the two names.

**Comment by dimitri-yatsenko** [comment]
The only problem that remains unresolved is issue #40. This is not a problem in python, because tables are created at the time of importing the module. In MATLAB, tables are created at the time of first use.  This causes an implicit commit during transactions, which is a particular problem for part relations because they might not be first used until they are called from inside their master's `makeTuples` call.

**Comment by dimitri-yatsenko** [comment]
One solution for #40 is to make the master aware of all its parts and make it instantiate all the parts before entering the `makeTuples` transaction.  What's the best syntax or mechanism for that?

**Comment by eywalker** [comment]
We might have to crawl the search path to discover all the classes that inherits from dj

**Comment by dimitri-yatsenko** [comment]
Crawling the search path could take seconds every time.  

One approach is to declare parts as properties of master. 

```matlab
classdef Master:
    properties(Constant)
        parts = [schema.Part1, schema.Part2]
    end
end
``` 

One problem is that matlab has a hard time clearing properties of classes.

**Comment by eywalker** [comment]
I would be strongly opposed to manipulating (constant) properties in Matlab - it' gets messy very quickly. Maybe we can maintain the list of parts in a separate file that gets read in? This sounds terrible because it'll be another separate file, but it will be cleaner update.

**Comment by dimitri-yatsenko** [comment]
Parts could also be listed in the table declaration.

**Comment by eywalker** [comment]
But if you modify it later, wouldn't that be a mess?

**Comment by dimitri-yatsenko** [comment]
We could list them in a string:

```matlab 
classdef Master < dj.Relvar
    properties(Constant)
        parts = 'Part1, Part2'
    end
end
```

**Comment by eywalker** [comment]
Actually I take that back - if we put that as part of the comment at the top, it's no different from my separate file with list solution.

**Comment by dimitri-yatsenko** [comment]
@eywalker  and I just spoke and came up with the following: 

1.  If a table declaration is attempted within a transaction, an error will be raised.
2. Master and part must be in the same schema (that's also true in Python)
2. Parts will be declared with `part of schema.Master` in the first line of the declaration. 
3. When `populate` gets called, it will examine all the .m files in the package, parse the declarations, and find all classes that declare themselves as part of the calling class.  It will then instantiate all the part classes and trigger their declarations.

**Comment by dimitri-yatsenko** [comment]
Fixed in #72 by introducing a the `dj.Part` class.

**Comment by eywalker** [comment]
Master and pard table concepts are now formally introduced into DataJoint MATLAB beginning in `v3.0.0` (PR #83). @dimitri-yatsenko Should we just close this issue now?
